### PR TITLE
Added SQL trace feature

### PR DESCRIPTION
**Pull request by boxed**

### CODE REVIEW

File: iommi/sql_trace.py [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+        else:
+            return getattr(self.cursor, attr)
+
+
+def make_debug_cursor(self, cursor):
+    return CursorDebugWrapper(cursor, self)
+
+
+BaseDatabaseWrapper.make_debug_cursor = make_debug_cursor
+# BaseDatabaseWrapper.make_cursor = make_debug_cursor
```
Comment (jlubcke) [inline iommi/sql_trace.py:401]: ?

File: iommi/sql_trace.py [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+def set_sql_debug(new_state, validate=True):
+    if new_state == 'None':
+        new_state = None
+    if validate:
+        assert new_state is None or new_state in SQL_DEBUG_LEVELS
+    setattr(state, 'sql_debug', new_state)
+
+
+def get_sql_debug():
+    """ :rtype: bool """
```
Comment (jlubcke) [inline iommi/sql_trace.py:191]: This docstring is old cruft I think

File: iommi/sql_trace.py [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+SQL_DEBUG_LEVEL_ALL = 'all'
+SQL_DEBUG_LEVEL_ALL_WITH_STACKS = 'stacks'
+SQL_DEBUG_LEVEL_WORST = 'worst'
+
+
+SQL_DEBUG_LEVELS = {
+    SQL_DEBUG_LEVEL_ALL,
+    SQL_DEBUG_LEVEL_ALL_WITH_STACKS,
+    SQL_DEBUG_LEVEL_WORST,
+    False,
```
Comment (jlubcke) [inline iommi/sql_trace.py:220]: A bit weird to have a bool among the strings. Would not a setting to None be more natural? (So the field is optional.)

File: iommi/sql_trace.py [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+    Context manager to temporarily suspend sql logging.
+
+    This is useful inside the sql debug implementation to avoid infinite recursion.
+    """
+    old_state = get_sql_debug()
+    set_sql_debug(False)
+    yield
+    set_sql_debug(old_state)
+
+
```
Comment (jlubcke) [inline iommi/sql_trace.py:210]: I think the options would be easier to find if they were at the top of the file.

File: iommi/sql_trace.py [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+    return sum(x['duration'] for x in log)
+
+
+def sql_debug_format_stack_trace(frame):
+    lines = traceback.format_stack(frame)
+
+    base_path = os.path.abspath(os.path.join(settings.BASE_DIR, '..')) + "/"
+    msg = []
+    skip_template_code = False
+    for line in itertools.dropwhile(lambda l: ('/lib/python' in l
```
Comment (jlubcke) [inline iommi/sql_trace.py:281]: newline before the string

File: iommi/sql_trace.py (hunk 1 above)
Comment (boxed) [reply iommi/sql_trace.py:401]: Fixed.

File: iommi/sql_trace.py (hunk 3 above)
Comment (boxed) [reply iommi/sql_trace.py:220]: Fixed

File: iommi/sql_trace.py (hunk 4 above)
Comment (boxed) [reply iommi/sql_trace.py:210]: Fixed.

File: iommi/sql_trace.py (hunk 5 above)
Comment (boxed) [reply iommi/sql_trace.py:281]: That didn't improve it. I moved it to a function instead.

**Pull request merged**